### PR TITLE
add option to enable static runtime builds for qt

### PR DESCRIPTION
**Pull request by Dargun**

### CODE REVIEW

**Comment by ruslo** [comment]
Please check `_DLL` macro in toolchain-id to set this option automatically. See my comment:
* https://github.com/ruslo/hunter/issues/1190#issuecomment-343739373

Example:
* https://github.com/ruslo/hunter/blob/77afb6e377572da9994a326cb6a6a9f9d77a2bb3/cmake/projects/OpenSSL/hunter.cmake#L335-L340

Since build scheme changed please update `PACKAGE_INTERNAL_DEPS_ID`:
* https://docs.hunter.sh/en/latest/reference/user-modules/hunter_download.html

**Comment by Dargun** [comment]
[Right,](https://github.com/ruslo/hunter/issues/1190#issuecomment-343739373)  if you want to compile static you have to compile all packages with /MT.
That works so far if you include include("${CMAKE_CURRENT_LIST_DIR}/flags/vs-mt.cmake") in the toolchain.
problem currently is that qt build pipeline is completly ignoring all this configuration from toolchain.
The only parameter that can be used to configure qt from outside is BUILD_SHARED_LIBS.
Maybe i'am wrong but thats what i have seen on my buildtests and scripts.
All magic with qt and project generation happens with this qtconfig tool with qt spezific flags like -static;-static-runtime. Thats the current palace where all project and compiler configrutations happen.
And the CMAKE param from hunter config "BUILD_SHARED_LIBS" is only adding "-static" to this list of configuration params for qtconfig usage. For static runtime builds we need to add additionally "-static-runtime".

**Comment by ruslo** [comment]
> Right, if you want to compile static you have to compile all packages with /MT.

By "static" you mean what? I can compile static library **without** `/MT` and link it to executable. Will work fine.

> currently is that qt build pipeline is completly ignoring all this configuration from toolchain

Because Qt use custom build scheme. Somebody have to teach it to respect everything.

**But** my suggestion is to use `hunter_check_toolchain_definition` in `hunter.cmake` (see example above) which is logically called from user's CMakeLists.txt in the same way as `include(/.../hunter.cmake)`. It means we respect toolchain parameters.

> And the CMAKE param from hunter config "BUILD_SHARED_LIBS" is only adding "-static" to this list of configuration params for qtconfig usage. For static runtime builds we need to add additionally "-static-runtime".

Building static library doesn't imply `/MT`. That's why Qt has two different options `-static` and `-static-runtime`. From Hunter point of view static/shared library controlled by `BUILD_SHARED_LIBS` and static/runtime runtime `/MT` vs. `/MD` are a globa flags => hence should be set in toolchain.

File: cmake/projects/Qt/schemes/url_sha1_qt.cmake.in [hunk 1]
```diff
@@ -215,6 +216,10 @@ if(BUILD_SHARED_LIBS)
   list(APPEND configure_opts "-shared")
 else()
   list(APPEND configure_opts "-static")
+  hunter_check_toolchain_definition(NAME "_MT" DEFINED _hunter_mt)
```
Comment (ruslo) [inline cmake/projects/Qt/schemes/url_sha1_qt.cmake.in]: `_MT` defined by both `/MT` and `/MD`. As far as I understand we need to check `_DLL`:
* https://github.com/ruslo/hunter/issues/1190#issuecomment-343739373
* https://github.com/ruslo/hunter/pull/1203#issuecomment-344561420

Comment (Dargun) [reply cmake/projects/Qt/schemes/url_sha1_qt.cmake.in]: you right, i have to check both toolchain definitions.

**Comment by Dargun** [comment]
ok, yea take ages to compile under windows :(

File: cmake/projects/Qt/schemes/url_sha1_qt.cmake.in [hunk 2]
```diff
@@ -215,6 +216,10 @@ if(BUILD_SHARED_LIBS)
   list(APPEND configure_opts "-shared")
 else()
   list(APPEND configure_opts "-static")
+  hunter_check_toolchain_definition(NAME "_DLL" DEFINED _hunter_vs_md)
```
Comment (ruslo) [inline cmake/projects/Qt/schemes/url_sha1_qt.cmake.in]: Not working. Some variables are not ready at this point. It's easier to call `hunter_check_toolchain_definition` in `hunter.cmake`.

To test you change locally:
* https://docs.hunter.sh/en/latest/creating-new/create/cmake.html#testing-locally

File: cmake/projects/Qt/schemes/url_sha1_qt.cmake.in [hunk 3]
```diff
@@ -215,6 +215,9 @@ if(BUILD_SHARED_LIBS)
   list(APPEND configure_opts "-shared")
 else()
   list(APPEND configure_opts "-static")
+  if(NOT QT_BUILD_DYNAMIC_VSRUNTIME)
+    list(APPEND configure_opts "-static-runtime")
```
Comment (ruslo) [inline cmake/projects/Qt/schemes/url_sha1_qt.cmake.in]: Linux build failed with message: `-static-runtime: invalid command-line switch`. I guess what I was afraid [will happen with Boost](https://github.com/ruslo/hunter/pull/1224#discussion_r153008429) happened with Qt :) We need `if(MSVC)` protection here.

**Comment by ruslo** [comment]
Tests passed:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.2879
* https://travis-ci.org/ingenue/hunter/builds/307834437?utm_source=github_status&utm_medium=notification

**Comment by ruslo** [comment]
Released: https://github.com/ruslo/hunter/releases/tag/v0.19.172

**Pull request merged**